SimTKcommon/src/Pathname.cpp

```diff
@@ -218,13 +220,187 @@ void Pathname::deconstructPathname( const string&   name,
-    if (isAbsolutePath) {
+    if (dontApplySearchPath) {
         if (!drive.empty())
             directory = drive + ":";
         directory += "/";
     }
 
-    for (int i = (int)segmentsInReverse.size()-1; i >= 0; --i)
+    for (int i = (int)segmentsInReverse.size() - 1; i >= 0; --i)
+        directory += segmentsInReverse[i] + "/";
+
+    // Fix the slashes.
+    makeNativeSlashesInPlace(directory);
+
+    // If there is a .extension, strip it off.
+    string::size_type lastDot = fileName.rfind('.');
+    if (lastDot != string::npos) {
+        extension = fileName.substr(lastDot);
+        fileName.erase(lastDot);
+    }
+}
+
+void Pathname::deconstructPathnameUsingSpecifiedWorkingDirectory(const std::string& swd,
+                                                                 const std::string& path,
+                                                                 std::string& directory,
+                                                                 std::string& fileName,
+                                                                 std::string& extension)
+{
+    directory.erase(); fileName.erase(); extension.erase();
+    string pathdrive, swddrive, finaldrive;
+    String processed;
+
+    // Remove all the white space and make all the slashes be forward ones.
+    // (For Windows they'll be changed to backslashes later.)
+    String pathcleaned = String::trimWhiteSpace(path).replaceAllChar('\\', '/');
+    if (pathcleaned.empty())
+        return; // path consisted only of white space
+    removeDriveInPlace(pathcleaned, pathdrive);
+
+    String swdcleaned = String::trimWhiteSpace(swd).replaceAllChar('\\', '/');
+    
+    // If swd was empty, then just call deconstructPathname() on the path.
+    if (swdcleaned.empty()) {
+        bool dontApplySearchPath;
+        deconstructPathname(path, dontApplySearchPath, directory, fileName, extension);
+        if (!dontApplySearchPath) 
+            directory = getCurrentWorkingDirectory() + directory;
+        return;
+    }
+
+    // If it wasn't empty, then add a trailing "/" if necessary.
+    if (swdcleaned[swdcleaned.size() - 1] != '/')
+        swdcleaned += '/';
+    removeDriveInPlace(swdcleaned, swddrive);
+
+    /* CHECK IF SWD IS NEEDED BY CHECKING IF PATH IS ROOT-RELATIVE */
+    // "@/" is a definitely a root-relative path. Just use deconstructPathname().
+    if (pathcleaned.substr(0, 2) == "@/") {
+        bool dontApplySearchPath;
+        deconstructPathname(path, dontApplySearchPath, directory, fileName, extension);
+        if (!dontApplySearchPath) 
+            directory = getCurrentWorkingDirectory() + directory;
+        return;
+    }
+
+    // Check if pathcleaned is a root-relative path. For Windows, this means
+    // that path gave a drive. For not-Windows, no drives will exist.
+    if (pathcleaned.substr(0, 1) == "/") {
+        if (getCurrentDriveLetter().empty() || !pathdrive.empty()) {
```

Comment:
Currently I think there is no directly platform-dependent code in this method. Instead everything is hidden by platform-dependent methods that understand how to interpret the path characters so that "C:" just comes in as ordinary file characters on Linux. It would be good to keep it that way if possible. Testing whether `getCurrentDriveLetter().empty()` is really a hidden platform dependency; those should be done with `#ifdef WIN32` and localized with platform dependencies elsewhere. In contrast testing `pathdrive.empty()` is not a platform dependency since it could be empty anywhere.